### PR TITLE
LB link removal removes service from lb

### LINKED ISSUE #1622: Load Balancer's haproxy.cfg not being updated after deploy
I recently tried to stick a varnish cache image in between my load balancer and my assets server. I was seeing performance improvements hitting varnish directly, but none at the load balancer. Then I reviewed the config, and realized it was only partially being updated. Here's an illustration of before and after in my `kontena.yml`.

This is before, it was balancing directly to the assets server:

```yaml
---
version: '2'
name: rails_app
stack: corylogan/mystack
services:
  load_balancer:
    image: kontena/lb:latest
    deploy:
      strategy: daemon
    ports:
      - 80:80
      - 443:443
    secrets:
      - secret: SSL_CERTS
        name: SSL_CERTS
        type: env

  minio:
    stateful: true
    image: corylogan/minio-distributed-kontena
    instances: 4
    ports:
      - 9999:9999
    command: /entrypoint.sh
    volumes:
      - /export
    environment:
      - KONTENA_LB_MODE=http
      - KONTENA_LB_BALANCE=roundrobin
      - KONTENA_LB_INTERNAL_PORT=9000
      - KONTENA_LB_EXTERNAL_PORT=9000
      - KONTENA_LB_VIRTUAL_HOSTS=minio.xyz.co
      - KONTENA_LB_CUSTOM_SETTINGS=redirect scheme https if !{ ssl_fc }
    links:
      - load_balancer
```

Then adding varnish in:

```yaml
---
version: '2'
name: rails_app
stack: corylogan/mystack
services:
  load_balancer:
    image: kontena/lb:latest
    deploy:
      strategy: daemon
    ports:
      - 80:80
      - 443:443
    secrets:
      - secret: SSL_CERTS
        name: SSL_CERTS
        type: env

  varnish:
    image: eeacms/varnish
    instances: 4
    ports:
      - 9998:9998
    environment:
      - KONTENA_LB_MODE=http
      - KONTENA_LB_BALANCE=roundrobin
      - KONTENA_LB_EXTERNAL_PORT=443
      - KONTENA_LB_INTERNAL_PORT=6081
      - KONTENA_LB_VIRTUAL_HOSTS=minio.xyz.co
      - KONTENA_LB_CUSTOM_SETTINGS=redirect scheme https if !{ ssl_fc }
      - BACKENDS=minio
      - BACKENDS_PORT=9000
      - DNS_ENABLED=true
      - CACHE_SIZE=64M
      - BACKENDS_PROBE_URL=/minio/logo.svg
    links:
      - minio
      - load_balancer

  minio:
    stateful: true
    image: corylogan/minio-distributed-kontena
    instances: 4
    ports:
      - 9999:9999
    command: /entrypoint.sh
    volumes:
      - /export

```

I've removed some stuff to try to make the point more clear, so please keep that in mind if i messed something up in the config.

The trouble is in the haproxy config that comes out. I'm cutting out the same parts here. Commenting in the important parts.

```
global
    daemon                  off
    pidfile                 /var/run/haproxy.pid
    user                    haproxy
    group                   haproxy
    maxconn                 100000
    log                     127.0.0.1 local1 info
    tune.ssl.default-dh-param 2048
    ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
defaults
    log                     global
    mode                    http
    option                  splice-auto
    option                  http-keep-alive
    option                  redispatch
    option                  httplog
    retries                 3
    timeout http-request    5s
    timeout queue           1m
    timeout connect         5s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
listen http-in
    bind *:80
    http-request replace-header Host (.*):.* \1
    bind *:443 ssl crt /etc/haproxy/certs/ no-sslv3
    reqadd X-Forwarded-Proto:\ https if { ssl_fc }
    reqadd X-Forwarded-Port:\ 443 if { ssl_fc }
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http

# here we see that the minio domain is still present for the assets host

          acl host_rails_app-minio hdr(host) -i minio.xyz.co

# And here we see that the  new acl has been added for varnish and the same domain

          acl host_rails_app-varnish hdr(host) -i minio.xyz.co
          acl host_rails_app-web hdr(host) -i kontena.xyz.co
          acl host_rails_app-web hdr(host) -i load_balancer.rails_app.production.kontena.local

# This next line should be gone as well

        use_backend rails_app-minio if host_rails_app-minio
        use_backend rails_app-varnish if host_rails_app-varnish
        use_backend rails_app-web if host_rails_app-web

# This backend shouldn't be here as i've removed the link and all LB ENV vars.

backend rails_app-minio
    option                  forwardfor
        redirect scheme https if !{ ssl_fc }
    balance                 roundrobin
    server rails_app-minio-1 10.81.128.20:9000 check 
    server rails_app-minio-2 10.81.128.8:9000 check 
    server rails_app-minio-3 10.81.128.1:9000 check 
    server rails_app-minio-4 10.81.128.104:9000 check 
backend rails_app-varnish
    option                  forwardfor
        redirect scheme https if !{ ssl_fc }
    balance                 roundrobin
    server rails_app-varnish-1 10.81.128.43:6081 check 
    server rails_app-varnish-2 10.81.128.30:6081 check 
    server rails_app-varnish-3 10.81.128.81:6081 check 
    server rails_app-varnish-4 10.81.128.118:6081 check 
backend rails_app-web
    option                  forwardfor
        redirect scheme https if !{ ssl_fc }
    balance                 roundrobin
    server rails_app-web-1 10.81.128.45:9292 check 
    server rails_app-web-2 10.81.128.106:9292 check 
    server rails_app-web-3 10.81.128.64:9292 check 
    server rails_app-web-4 10.81.128.66:9292 check 
    server rails_app-web-5 10.81.128.98:9292 check 
listen stats
    mode http
    bind 0.0.0.0:1000
    stats enable
    stats uri /
    stats refresh 10s
    stats show-node
    stats show-legends
    # if authentication is wanted
    acl auth_ok http_auth(stats-auth)
    http-request auth unless auth_ok
userlist stats-auth
    user stats insecure-password secret
```

Interestingly `kon stack show stack rails_app` yields the configuration I'm expecting: 

```
rails_app:
  state: running
  created_at: 2017-01-04T10:21:50.552Z
  updated_at: 2017-01-04T10:21:50.552Z
  version: 2
  expose: -
  services:
    web:
      image: registry.production.kontena.local/xyz-production:lates
      status: running
      revision: 8
      stateful: no
      scaling: 5
      strategy: ha
      deploy_opts:
        min_health: 0.8
      wait_for_port: 9292
      dns: web.rails_app.production.kontena.local
      links:
        - load_balancer => production/rails_app/load_balancer
    load_balancer:
      image: kontena/lb:latest
      status: running
      revision: 2
      stateful: no
      scaling: 1
      strategy: daemon
      deploy_opts:
        min_health: 0.8
      dns: load_balancer.rails_app.production.kontena.local
      ports:
        - 80:80/tcp
        - 443:443/tcp
    minio:
      image: corylogan/minio-distributed-kontena:latest
      status: running
      revision: 2
      stateful: yes
      scaling: 4
      strategy: ha
      deploy_opts:
        min_health: 0.8
      dns: minio.rails_app.production.kontena.local
      cmd: /entrypoint.sh
      ports:
        - 9999:9999/tcp
    varnish:
      image: eeacms/varnish:latest
      status: running
      revision: 3
      stateful: no
      scaling: 4
      strategy: ha
      deploy_opts:
        min_health: 0.8
      dns: varnish.rails_app.production.kontena.local
      ports:
        - 9998:9998/tcp
      links:
        - minio => production/rails_app/minio
        - load_balancer => production/rails_app/load_balancer
```

As you can see above there is no link to the load balancer from minio, and there is for varnish.

Thanks for all the help folks!

### PR DESCRIPTION
**Pull request by jnummelin**
When lb link is removed from the service, load balancer configuration was left behind.

This PR fixes that by issuing total service removal on all lb's when service is deployed with no links to loadbalancers.
fixes #1622 

```
$ k service create --link lb -e KONTENA_LB_VIRTUAL_PATH=/ nginx nginx && k service deploy nginx 
...
$ k etcd ls --recursive /kontena/haproxy
/kontena/haproxy
/kontena/haproxy/lb
/kontena/haproxy/lb/aliases
/kontena/haproxy/lb/certs
/kontena/haproxy/lb/services
/kontena/haproxy/lb/services/nginx
/kontena/haproxy/lb/services/nginx/upstreams
/kontena/haproxy/lb/services/nginx/upstreams/null-nginx-1
/kontena/haproxy/lb/services/nginx/balance
/kontena/haproxy/lb/services/nginx/virtual_path
/kontena/haproxy/lb/tcp-services
```
After unlinking and deployment:
```
$ k etcd ls --recursive /kontena/haproxy
/kontena/haproxy
/kontena/haproxy/lb
/kontena/haproxy/lb/aliases
/kontena/haproxy/lb/certs
/kontena/haproxy/lb/services
/kontena/haproxy/lb/tcp-services
```

### CODE REVIEW

File: agent/lib/kontena/service_pods/creator.rb [hunk 1]
```diff
@@ -49,6 +49,8 @@ def perform
         debug "container created: #{service_pod.name}"
         if service_container.load_balanced? && service_container.instance_number == 1
           Celluloid::Notifications.publish('lb:ensure_config', service_container)
+        else
+          Celluloid::Notifications.publish('lb:remove_service', service_container)
```
Comment (jakolehm) [inline agent/lib/kontena/service_pods/creator.rb]: Should we publish this also only with first instance?

Comment (SpComb) [inline agent/lib/kontena/service_pods/creator.rb]: Won't this nuke the service for anything scaled to more than 1 instance?

File: agent/lib/kontena/load_balancers/configurer.rb [hunk 2]
```diff
@@ -90,6 +95,31 @@ def remove_config(container)
       error "#{exc.class.name}: #{exc.message}"
     end
 
+
+    # @param [Docker::Container] container
+    def remove_service(container)
+      service_name = container.service_name_for_lb
+      lsdir(ETCD_PREFIX).each do |key|
+        if key_exists?("#{key}/services")
```
Comment (SpComb) [inline agent/lib/kontena/load_balancers/configurer.rb]: or `tcp-services`?

File: agent/lib/kontena/service_pods/creator.rb (hunk 1 above)
Comment (jnummelin) [inline agent/lib/kontena/service_pods/creator.rb]: Good point. Will refactor this.

**Comment by jnummelin** [comment]
Updated the `service_uptodate?` logic to look into labels to see if there are lb changes. This triggers now the service lb removal only on first instance of a service.

**Pull request merged**